File: iosApp/iosApp/ViewModels/SearchViewModel.swift
//
//  SearchViewModel.swift
//  iosApp
//
//  Created by esimon on 9/12/24.
//  Copyright © 2024 MBTA. All rights reserved.
//

@_spi(Experimental) import MapboxMaps
import Shared

class SearchViewModel: ObservableObject {
    enum ResultsState: Equatable {
        case loading
        case recentStops(stops: [Result])
        case results(stops: [Result], routes: [RouteResult])
        case error

        func isEmpty(includeRoutes: Bool) -> Bool {
            switch self {
            case .loading: false
            case .recentStops: false
            case let .results(stops: stops, routes: routes): stops.isEmpty && (!includeRoutes || routes.isEmpty)
            case .error: false
            }
        }
    }

    struct Result: Identifiable, Equatable {
        let id: String
        let isStation: Bool
        let name: String
        let routePills: [RoutePillSpec]
    }

    @Published var resultsState: ResultsState?

    private let visitHistoryUsecase: VisitHistoryUsecase
    private let searchResultsRepository: ISearchResultRepository
    private let globalRepository: IGlobalRepository

    private let analytics: Analytics

    private var globalResponse: GlobalResponse?
    private var latestVisits: [Result]?
    private var fetchResultsTask: Task<Void, Never>?

    init(
        globalRepository: IGlobalRepository = RepositoryDI().global,
        visitHistoryUsecase: VisitHistoryUsecase = UsecaseDI().visitHistoryUsecase,
        searchResultsRepository: ISearchResultRepository = RepositoryDI().searchResults,
        analytics: Analytics = AnalyticsProvider.shared
    ) {
        self.globalRepository = globalRepository
        self.visitHistoryUsecase = visitHistoryUsecase
        self.searchResultsRepository = searchResultsRepository
        self.analytics = analytics
    }

    func getStopFor(id: String) -> Stop? {
        globalResponse?.getStop(stopId: id)
    }

    func determineStateFor(query: String) {
        analytics.performedSearch(query: query)
        if query.isEmpty {
            fetchResultsTask?.cancel()
            resultsState = nil
            showRecentStops()
        } else {
            fetchResultsTask?.cancel()
            fetchResultsTask = Task { await loadResults(query: query) }
        }
    }

    func loadGlobalDataAndHistory() async {
        await getGlobalData()
    }

    @MainActor
    func activateGlobalListener() async {
        for await globalData in globalRepository.state {
            globalResponse = globalData
            Task {
                await loadVisitHistory()
            }
        }
    }

    private func getGlobalData() async {
        Task(priority: .high) {
            await activateGlobalListener()
        }
        do {
            let result = try await onEnum(of: globalRepository.getGlobalData())
            if case let .error(error) = result { debugPrint(error) }
        } catch {}
    }

    private func loadVisitHistory() async {
        do {
            let latestVisits = try await visitHistoryUsecase.getLatestVisits()
                .compactMap { visit in
                    switch onEnum(of: visit) {
                    case let .stopVisit(stopVisit):
                        mapStopIdToResult(id: stopVisit.stopId)
                    }
                }
            await MainActor.run { [latestVisits] in
                self.latestVisits = latestVisits
                if resultsState == nil {
                    showRecentStops()
                }
            }
        } catch {}
    }

    private func loadResults(query: String) async {
        do {
            var resultsState: ResultsState? = nil
            resultsState = .loading
            switch try await onEnum(of: searchResultsRepository.getSearchResults(query: query)) {
            case let .ok(result):
                let results = result.data
                resultsState = .results(
                    stops: results.stops.compactMap { mapStopIdToResult(id: $0.id) },
                    routes: results.routes
                )
            case nil: resultsState = .error
            case let .error(error):
                resultsState = .error
                debugPrint(error)
            }
            guard !Task.isCancelled else { return }
            await MainActor.run { [resultsState] in
                self.resultsState = resultsState
            }
        } catch {}
    }

    private func showRecentStops() {
        guard let latestVisits, !latestVisits.isEmpty else { return }
        resultsState = .recentStops(stops: latestVisits)
    }

    private func mapStopIdToResult(id: String) -> Result? {
        guard let globalResponse,
              let stop = globalResponse.getStop(stopId: id)
        else { return nil }
        let isStation = stop.locationType == .station
        let routes = globalResponse.getTypicalRoutesFor(stopId: id)
        let routePills: [RoutePillSpec] = routes
            .sorted(by: { $0.sortOrder < $1.sortOrder })
            .map { route -> RoutePillSpec in
                let line: Line? = globalResponse.getLine(lineId: route.lineId)
                let context: RoutePillSpec.Context = isStation ? .searchStation : .default
                return RoutePillSpec(route: route,
                                     line: line,
                                     type: .flexCompact,
                                     context: context,
                                     contentDescription: stopRouteContentDescription(
                                         isStation: isStation,
                                         route: route
                                     ))
            }

        return Result(
            id: stop.id,
            isStation: isStation,
            name: stop.name,
            routePills: routePills.removingDuplicates()
        )
    }

    private func stopRouteContentDescription(isStation: Bool, route: Route) -> String {
        if silverRoutes.contains(route.id), isStation {
            let routeName = "Silver Line"
            return "\(routeName) \(route.type.typeText(isOnly: false))"

        } else if route.type == .commuterRail, isStation {
            let routeName = "Commuter Rail"
            return "\(routeName) \(route.type.typeText(isOnly: false))"
        } else if route.type == .bus, isStation {
            return route.type.typeText(isOnly: false)
        } else {
            return "\(route.label) \(route.type.typeText(isOnly: true))"
        }
    }
}

File: shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/viewModel/SearchViewModel.kt
package com.mbta.tid.mbta_app.viewModel

import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import co.touchlab.skie.configuration.annotations.DefaultArgumentInterop
import com.mbta.tid.mbta_app.analytics.Analytics
import com.mbta.tid.mbta_app.history.Visit
import com.mbta.tid.mbta_app.model.Line
import com.mbta.tid.mbta_app.model.LocationType
import com.mbta.tid.mbta_app.model.Route
import com.mbta.tid.mbta_app.model.RoutePillSpec
import com.mbta.tid.mbta_app.model.RouteType
import com.mbta.tid.mbta_app.model.response.ApiResult
import com.mbta.tid.mbta_app.model.response.GlobalResponse
import com.mbta.tid.mbta_app.model.silverRoutes
import com.mbta.tid.mbta_app.repositories.IGlobalRepository
import com.mbta.tid.mbta_app.repositories.ISearchResultRepository
import com.mbta.tid.mbta_app.repositories.ISentryRepository
import com.mbta.tid.mbta_app.usecases.VisitHistoryUsecase
import kotlin.jvm.JvmName
import kotlin.time.Duration.Companion.seconds
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.IO
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.withContext

private fun stopRouteContentDescription(
    isStation: Boolean,
    route: Route,
): RoutePillSpec.ContentDescription.StopSearchResultRoute {
    if (silverRoutes.contains(route.id) && isStation) {
        val routeName = "Silver Line"
        return RoutePillSpec.ContentDescription.StopSearchResultRoute(
            routeName,
            route.type,
            isOnly = false,
        )
    } else if (route.type == RouteType.COMMUTER_RAIL && isStation) {
        val routeName = "Commuter Rail"
        return RoutePillSpec.ContentDescription.StopSearchResultRoute(
            routeName,
            route.type,
            isOnly = false,
        )
    } else if (route.type == RouteType.BUS && isStation) {
        return RoutePillSpec.ContentDescription.StopSearchResultRoute(
            routeName = null,
            route.type,
            isOnly = false,
        )
    } else {
        return RoutePillSpec.ContentDescription.StopSearchResultRoute(
            route.label,
            route.type,
            isOnly = true,
        )
    }
}

public interface ISearchViewModel {

    public val models: StateFlow<SearchViewModel.State>

    public fun setQuery(query: String)

    public fun refreshHistory()
}

public class SearchViewModel(
    private val analytics: Analytics,
    private val globalRepository: IGlobalRepository,
    private val searchResultRepository: ISearchResultRepository,
    private val sentryRepository: ISentryRepository,
    private val visitHistoryUsecase: VisitHistoryUsecase,
) : MoleculeViewModel<SearchViewModel.Event, SearchViewModel.State>(), ISearchViewModel {
    public sealed interface Event {
        public data object RefreshHistory : Event
    }

    public sealed class State {
        public data object Loading : State()

        public data class RecentStops(val stops: List<StopResult>) : State()

        public data class Results(val stops: List<StopResult>, val routes: List<RouteResult>) :
            State()

        public data object Error : State()

        public fun isEmpty(includeRoutes: Boolean): Boolean =
            when (this) {
                Loading -> false
                is RecentStops -> false
                is Results -> stops.isEmpty() && !(includeRoutes && routes.isNotEmpty())
                Error -> false
            }
    }

    public data class StopResult(
        val id: String,
        val isStation: Boolean,
        val name: String,
        val routePills: List<RoutePillSpec>,
    ) {
        internal companion object {
            fun forStop(stopId: String, globalData: GlobalResponse?): StopResult? {
                val stop = globalData?.getStop(stopId) ?: return null
                val isStation = stop.locationType == LocationType.STATION
                val routes = globalData.getTypicalRoutesFor(stopId)
                val routePills =
                    routes.sorted().map { route ->
                        val line: Line? = globalData.getLine(route.lineId)
                        val context: RoutePillSpec.Context =
                            if (isStation) RoutePillSpec.Context.SearchStation
                            else RoutePillSpec.Context.Default
                        RoutePillSpec(
                            route,
                            line,
                            RoutePillSpec.Type.FlexCompact,
                            RoutePillSpec.Height.Small,
                            context,
                            stopRouteContentDescription(isStation, route),
                        )
                    }

                return StopResult(
                    id = stop.id,
                    isStation = isStation,
                    name = stop.name,
                    routePills = routePills.distinct(),
                )
            }
        }
    }

    public data class RouteResult(val id: String, val name: String, val routePill: RoutePillSpec) {
        public companion object {
            public fun forLineOrRoute(objectId: String, globalData: GlobalResponse?): RouteResult? {
                val route = globalData?.getRoute(objectId)
                val line = globalData?.getLine(objectId)

                val routePillSpec =
                    RoutePillSpec(
                        route,
                        line,
                        RoutePillSpec.Type.Fixed,
                        context = RoutePillSpec.Context.Default,
                    )

                return RouteResult(
                    id = objectId,
                    name = line?.longName ?: route?.longName ?: return null,
                    routePill = routePillSpec,
                )
            }
        }
    }

    @set:JvmName("setQueryState") private var query by mutableStateOf("")

    @Composable
    override fun runLogic(): State {
        val globalData by globalRepository.state.collectAsState()
        var latestVisits by remember { mutableStateOf<List<Visit>?>(null) }
        var state by remember { mutableStateOf<State>(State.Loading) }

        LaunchedEffect(null) { globalRepository.getGlobalData() }

        LaunchedEffect(null) { latestVisits = visitHistoryUsecase.getLatestVisits() }

        EventSink(eventHandlingTimeout = 2.seconds, sentryRepository = sentryRepository) { event ->
            when (event) {
                Event.RefreshHistory -> latestVisits = visitHistoryUsecase.getLatestVisits()
            }
        }

        LaunchedEffect(query, latestVisits, globalData) {
            analytics.performedSearch(query)
            if (query.isNotEmpty()) {
                // explicitly not in dependencies to avoid thrashing every frame
                if (state is State.RecentStops) state = State.Loading
                withContext(Dispatchers.IO) {
                    delay(500)
                    when (val data = searchResultRepository.getSearchResults(query)) {
                        is ApiResult.Ok ->
                            state =
                                State.Results(
                                    data.data.stops.mapNotNull {
                                        StopResult.forStop(it.id, globalData)
                                    },
                                    data.data.routes.mapNotNull {
                                        RouteResult.forLineOrRoute(
                                            when {
                                                it.id == "Green" -> "line-Green"
                                                else -> it.id
                                            },
                                            globalData,
                                        )
                                    },
                                )
                        is ApiResult.Error -> state = State.Error
                        null -> {}
                    }
                }
            } else {
                val latestVisitResults =
                    latestVisits
                        ?.filterIsInstance<Visit.StopVisit>()
                        ?.mapNotNull { StopResult.forStop(it.stopId, globalData) }
                        ?.takeIf { it.isNotEmpty() }
                state =
                    if (latestVisitResults != null) State.RecentStops(latestVisitResults)
                    else State.Loading
            }
        }

        return state
    }

    override val models: StateFlow<State>
        get() = internalModels

    override fun setQuery(query: String) {
        this.query = query
    }

    override fun refreshHistory(): Unit = fireEvent(Event.RefreshHistory)
}

public class MockSearchViewModel
@DefaultArgumentInterop.Enabled
constructor(initialState: SearchViewModel.State = SearchViewModel.State.Loading) :
    ISearchViewModel {
    public var onSetQuery: (String) -> Unit = {}
    internal var onRefreshHistory = {}
    override val models: MutableStateFlow<SearchViewModel.State> = MutableStateFlow(initialState)

    override fun setQuery(query: String) {
        onSetQuery(query)
    }

    override fun refreshHistory() {
        onRefreshHistory()
    }
}
